# Generate Mappability Filter

This repo contains a collection of scripts that are used to generate a mappability filter bed file from a given fasta file using Heng Li's seqbility tools (https://lh3lh3.users.sourceforge.net/snpable.shtml). 

`make_mappability_filter.sh` generates a mask fasta file using the `bwa` aligner. `bwa` is used for mapping, but any aligner can be used to generate the filter.

`convert_seqbility_fa_to_bed.py` generates a bed file from the mask fasta file

The `convert_seqbility_fa_to_bed.py` script is designed to work on human genetic data but `make_mappability_filter.sh` can be used for any fasta file.

Briefly, a mappability filter is generated by splitting a given fasta file (`-i`) into kmers of a specified length (`-k`) and aligning these kmers back to the genome allowing for up to one mismatch. A mappability mask fasta file is then generated based on the stringency parameter (`-r`) which specifies the amount of times a given kmer can be aligned back to the genome.

`r = 0.5` -- 50% of all possible 35mers overlapping a position do not find a match to any other position in the genome

`r = 0.99` -- 100% all possible 35mers overlapping a position do not find a match to any other position in the genome

## Requirements
`bwa`, `samtools` and `parallel` are required by `make_mappability_filter.sh`

`pandas` is required by the `convert_seqbility_fa_to_bed.py` script

A bwa index must exist for the specified fasta file.
